model-engine/model_engine_server/entrypoints/start_fastapi_server.py

```diff
@@ -3,28 +3,42 @@
 
 You can do this with `start-fastapi-server`.
 """
+
 import argparse
 import subprocess
 from typing import List
 
+# Uvicorn returns 503 instead of 429 when concurrency exceeds the limit
+# We'll autoscale at target concurrency of a much lower number (around 50), and this just makes sure we don't 503 with bursty traffic
+# We set this very high since model_engine_server/api/app.py sets a lower per-pod concurrency at which we start returning 429s
+CONCURRENCY_LIMIT = 10000
+
 
-def start_gunicorn_server(port: int, num_workers: int, debug: bool) -> None:
-    """Starts a GUnicorn server locally."""
+def start_uvicorn_server(port: int, debug: bool) -> None:
+    """Starts a Uvicorn server locally."""
     additional_args: List[str] = []
     if debug:
-        additional_args.extend(["--reload", "--timeout", "0"])
+        additional_args.extend(["--reload", "--timeout-graceful-shutdown", "0"])
     command = [
-        "gunicorn",
-        "--bind",
-        f"[::]:{port}",
-        "--timeout",
+        "uvicorn",
+        "--host",
+        "::",
+        "--port",
+        f"{port}",
+        "--timeout-graceful-shutdown",
         "60",
-        "--keep-alive",
+        "--timeout-keep-alive",
         "2",
-        "--worker-class",
-        "model_engine_server.api.worker.LaunchWorker",
+        # uvloop and httptools are both faster than their alternatives, but they are not compatible
+        # with Windows or PyPy.
+        "--loop",
+        "uvloop",
+        "--http",
+        "httptools",
+        "--limit-concurrency",
+        f"{CONCURRENCY_LIMIT}",
         "--workers",
-        f"{num_workers}",
+        "1",  # Let the Kubernetes deployment handle the number of pods
```

Comment:
Will this reduce the amount of traffic we can receive per-pod? Why not keep it at 4?

Reply:
this is to remove load balancing within pod

Reply:
are we increasing the number of pods by 4 to compensate?

Reply:
that's the initial plan

Reply:
tbh i think llm engine is overprovisioned